### PR TITLE
BulkCreateAPI, Signals, HyperlinkedModelSerializers, Serializer Validations

### PR DESCRIPTION
**Pull request by maneeshatejani1**
https://www.loom.com/share/4c60bf3e7c6e4ccfad645816f5ca258b

### CODE REVIEW

File: books/views.py [hunk 1]
```diff
@@ -56,41 +75,56 @@ def read_or_update(request, book_id):
...
+            return Response(serializer.data, status=status.HTTP_200_OK)
+        else:
+            return Response({"message": "You cannot update an object other than your own"},
+                            status=status.HTTP_400_BAD_REQUEST)
 
     def delete(self, request, pk, format=None):
-        get_object_or_404(Author, pk=pk).delete()
-        return Response(status=status.HTTP_204_NO_CONTENT)
+        author = get_object_or_404(Author, pk=pk)
+        if request.user == author.user:
```
Comment (tehreem-sadat) [inline books/views.py]: Try to explore and handle this through custom Django permission classes. 

```
Class IsSelf:
    def has_permission()
        ...
```
Django view should be clean. All validations should be handled in serializes and permissions should be handled through custom permission classes.

File: books/models.py [hunk 2]
```diff
@@ -33,3 +35,28 @@ class Book(models.Model):
...
+class CustomLog(models.Model):
+    created_at = models.DateTimeField(auto_now_add=True)
+    message = models.CharField(max_length=500)
+    performed_by = models.ForeignKey(get_user_model(), on_delete=models.CASCADE)
+
+    def __str__(self):
+        return self.message
+
+
+@receiver(post_delete, sender=Author)
```
Comment (tehreem-sadat) [inline books/models.py]: Can we please move all your signal handler in separate file "signals.py". The "models.py" should only contain database models.

File: books/serializers.py [hunk 3]
```diff
@@ -1,24 +1,74 @@
...
-        user = get_user_model().objects.get(id=user_id)
-        return Author.objects.create(short_name=short_name, area_of_interest=area_of_interest, user=user)
-
-    def update(self, instance, validated_data):
-        instance.short_name = validated_data.get('short_name', instance.short_name)
-        instance.area_of_interest = validated_data.get('area_of_interest', instance.area_of_interest)
-        instance.save()
-        return instance
+        books = validated_data.get('book_set', None)
+        if books is None:
```
Comment (tehreem-sadat) [inline books/serializers.py]: Try to simplify if-else statements. Lines which are repeated in both conditions should be written either before or after if-else region.

File: books/serializers.py [hunk 4]
```diff
@@ -1,24 +1,74 @@
...
+            author_created = Author.objects.create(**validated_data)
+            return author_created
+        else:
+            validated_data.pop('book_set')
```
Comment (tehreem-sadat) [inline books/serializers.py]: `validated_data.pop('book_set', None)`

So it won't give key-error and we can use it outside if-else.

File: books/serializers.py [hunk 5]
```diff
@@ -1,24 +1,74 @@
...
+            author_created = Author.objects.create(**validated_data)
+            for book in books:
+                book = Book.objects.create(**book, author=author_created)
+            return author_created
+
+
+# class AuthorSerializer(serializers.Serializer):
```
Comment (tehreem-sadat) [inline books/serializers.py]: remove commented section.

**Review by tehreem-sadat — CHANGES_REQUESTED** [review]
Just few minor changes required.

File: books/views.py (hunk 1 above)
Comment (maneeshatejani1) [reply books/views.py]: sure

Comment (maneeshatejani1) [reply books/views.py]: done

File: books/models.py (hunk 2 above)
Comment (maneeshatejani1) [reply books/models.py]: done

File: books/serializers.py (hunk 3 above)
Comment (maneeshatejani1) [reply books/serializers.py]: done

File: books/serializers.py (hunk 4 above)
Comment (maneeshatejani1) [reply books/serializers.py]: done

File: books/serializers.py (hunk 5 above)
Comment (maneeshatejani1) [reply books/serializers.py]: done

File: books/serializers.py [hunk 6]
```diff
@@ -1,24 +1,51 @@
...
-        user = get_user_model().objects.get(id=user_id)
-        return Author.objects.create(short_name=short_name, area_of_interest=area_of_interest, user=user)
-
-    def update(self, instance, validated_data):
-        instance.short_name = validated_data.get('short_name', instance.short_name)
-        instance.area_of_interest = validated_data.get('area_of_interest', instance.area_of_interest)
-        instance.save()
-        return instance
+        books = validated_data.get('book_set', None)
+        validated_data.pop('book_set', None)
```
Comment (tehreem-sadat) [inline books/serializers.py]: - Combine line 45 and 46 i.e.
```suggestion
        books = validated_data.pop('book_set', None)
```

File: books/signals.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+def log_entry_for_delete_author(sender, instance, *args, **kwargs):
+    message = "author instance with id {} deleted".format(instance.id)
+    CustomLog.objects.create(message=message, performed_by=instance.user)
+
+
+@receiver(post_save, sender=Author)
+def log_entry_for_create_or_update_author(sender, instance, created, **kwargs):
+    if created:
+        message = "new author instance created with id {}".format(instance.id)
+        CustomLog.objects.create(message=message, performed_by=instance.user)
```
Comment (tehreem-sadat) [inline books/signals.py]: Same statement in if-else, move this out of if-else.

**Review by tehreem-sadat — APPROVED** [review]

**Pull request merged**